kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -85,10 +102,62 @@ public void forwardRequest(ApiMessage message) {
             throw new AssertionError("Attempt to use forwardRequest with a non-request: " + name);
         }
 
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Forwarding request: {}", channelDescriptor(), decodedFrame);
+        }
         // TODO check we've not forwarded it already
         channelContext.write(decodedFrame, promise);
     }
 
+    @Override
+    public <T extends ApiMessage> ProxyFuture<T> sendRequest(short apiVersion, ApiMessage message) {
+        checkNotClosed();
+        short key = message.apiKey();
+        var apiKey = ApiKeys.forId(key);
+        short headerVersion = apiKey.requestHeaderVersion(apiVersion);
+        var header = new RequestHeaderData()
+                .setCorrelationId(0)
+                .setRequestApiKey(key)
+                .setRequestApiVersion(apiVersion);
+        if (headerVersion > 1) {
+            header.setClientId(filter.getClass().getSimpleName() + "@" + System.identityHashCode(filter));
+        }
+        boolean hasResponse = apiKey != ApiKeys.PRODUCE
+                || ((ProduceRequestData) message).acks() != 0;
+        var filterPromise = new ProxyPromiseImpl<T>();
+        var frame = DecodedRequestFrame.internalRequest(
+                apiVersion, 0, hasResponse,
+                filter, filterPromise, header, message);
+
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Sending request: {}", channelDescriptor(), frame);
+        }
+        ChannelPromise writePromise = outboundCtx.voidPromise();
+        // if (outboundCtx.channel().isWritable()) {
+        // outboundCtx.write(frame, writePromise);
+        // }
+        // else {
+        outboundCtx.writeAndFlush(frame, writePromise);
```

Comment:
@franz1981I guess we need some backpressure logic. But I'm not certain what would make sense just yet. Maybe we should always stop reading from the client if we're making an upstream request?